[vllm] ModuleNotFoundError: No module named 'vllm_xpu_kernels._C'

```
$ ./scripts/test-triton.sh --vllm
**** Installing pip dependencies ****
Requirement already satisfied: lit in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 2)) (18.1.8)
Requirement already satisfied: autopep8 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 3)) (2.3.2)
Requirement already satisfied: isort in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 4)) (6.0.1)
Requirement already satisfied: numpy in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 5)) (2.0.2)
Requirement already satisfied: pytest in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 6)) (8.4.1)
Requirement already satisfied: pytest-forked in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 7)) (1.6.0)
Requirement already satisfied: pytest-xdist in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 8)) (3.8.0)
Requirement already satisfied: scipy>=1.7.1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 9)) (1.15.3)
Requirement already satisfied: llnl-hatchet in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 10)) (2025.1.0)
Requirement already satisfied: expecttest in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 11)) (0.3.0)
Requirement already satisfied: matplotlib in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (3.10.3)
Requirement already satisfied: pandas<3.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 15)) (2.3.1)
Requirement already satisfied: tabulate in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 16)) (0.9.0)
Requirement already satisfied: pytest-rerunfailures in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 21)) (15.1)
Requirement already satisfied: pytest-skip>=0.2.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 22)) (0.2.0)
Requirement already satisfied: pytest-timeout in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 23)) (2.4.0)
Requirement already satisfied: setproctitle==1.3.5 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from -r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 26)) (1.3.5)
Requirement already satisfied: pycodestyle>=2.12.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from autopep8->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 3)) (2.14.0)
Requirement already satisfied: tomli in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from autopep8->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 3)) (2.2.1)
Requirement already satisfied: exceptiongroup>=1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 6)) (1.2.2)
Requirement already satisfied: iniconfig>=1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 6)) (2.1.0)
Requirement already satisfied: packaging>=20 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 6)) (25.0)
Requirement already satisfied: pluggy<2,>=1.5 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 6)) (1.6.0)
Requirement already satisfied: pygments>=2.7.2 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 6)) (2.18.0)
Requirement already satisfied: py in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest-forked->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 7)) (1.11.0)
Requirement already satisfied: execnet>=2.1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest-xdist->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 8)) (2.1.1)
Requirement already satisfied: pydot in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from llnl-hatchet->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 10)) (4.0.1)
Requirement already satisfied: PyYAML in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from llnl-hatchet->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 10)) (6.0.2)
Requirement already satisfied: textX>=3.0.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from llnl-hatchet->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 10)) (4.2.2)
Requirement already satisfied: multiprocess in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from llnl-hatchet->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 10)) (0.70.18)
Requirement already satisfied: caliper-reader in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from llnl-hatchet->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 10)) (0.4.1)
Requirement already satisfied: contourpy>=1.0.1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from matplotlib->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (1.3.2)
Requirement already satisfied: cycler>=0.10 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from matplotlib->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (0.12.1)
Requirement already satisfied: fonttools>=4.22.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from matplotlib->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (4.59.0)
Requirement already satisfied: kiwisolver>=1.3.1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from matplotlib->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (1.4.8)
Requirement already satisfied: pillow>=8 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from matplotlib->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (11.3.0)
Requirement already satisfied: pyparsing>=2.3.1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from matplotlib->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (3.2.0)
Requirement already satisfied: python-dateutil>=2.7 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from matplotlib->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (2.9.0)
Requirement already satisfied: pytz>=2020.1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pandas<3.0->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 15)) (2023.4)
Requirement already satisfied: tzdata>=2022.7 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pandas<3.0->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 15)) (2024.2)
Requirement already satisfied: six>=1.5 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 14)) (1.16.0)
Requirement already satisfied: Arpeggio>=2.0.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from textX>=3.0.0->llnl-hatchet->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 10)) (2.0.2)
Requirement already satisfied: dill>=0.4.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from multiprocess->llnl-hatchet->-r /home/jovyan/intel-xpu-backend-for-triton/scripts/requirements-test.txt (line 10)) (0.4.0)
**** Installing pytorch ****
***** Using pinned PyTorch commit 9fd8273ddc547524eb5927bd4e04b7c180c5ed27 by default. *****
**** PyTorch is already installed. Current commit is 9fd8273 ****
**** PyTorch is already installed and its current commit matches the pinned commit: 9fd8273ddc547524eb5927bd4e04b7c180c5ed27. ****
**** There is no need to build anything, exiting. ****
************************************************
******    Running VLLM Triton tests       ******
************************************************
************************************************
******    Installing VLLM                 ******
************************************************
Cloning into 'vllm'...
remote: Enumerating objects: 170840, done.
remote: Counting objects: 100% (81/81), done.
remote: Compressing objects: 100% (70/70), done.
remote: Total 170840 (delta 36), reused 13 (delta 11), pack-reused 170759 (from 4)
Receiving objects: 100% (170840/170840), 156.72 MiB | 35.85 MiB/s, done.
Resolving deltas: 100% (135183/135183), done.
Note: switching to '97a042f3bca53417de6405a248e3d11fca568e2c'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 97a042f3b Make the `httpx` logger less annoying when Transformers v5 is installed (#30480)
vllm                               0.10.2rc3.dev151+g2b8569703.d20251124.xpu /home/jovyan/intel-xpu-backend-for-triton/vllm
vllm-xpu-kernels                   0.0.1                                     /home/jovyan/intel-xpu-backend-for-triton/vllm-xpu-kernels
Requirement already satisfied: cachetools in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (5.5.2)
Requirement already satisfied: cbor2 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (5.7.1)
Requirement already satisfied: blake3 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (1.0.8)
Requirement already satisfied: pybase64 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (1.4.2)
Requirement already satisfied: openai_harmony in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (0.0.8)
Requirement already satisfied: tblib in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (3.2.2)
Requirement already satisfied: typing-extensions>=4.6.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from blake3) (4.15.0)
Requirement already satisfied: pydantic>=2.11.7 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from openai_harmony) (2.12.4)
Requirement already satisfied: annotated-types>=0.6.0 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pydantic>=2.11.7->openai_harmony) (0.7.0)
Requirement already satisfied: pydantic-core==2.41.5 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pydantic>=2.11.7->openai_harmony) (2.41.5)
Requirement already satisfied: typing-inspection>=0.4.2 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pydantic>=2.11.7->openai_harmony) (0.4.2)
Requirement already satisfied: pytest in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (8.4.1)
Requirement already satisfied: pytest-cov in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (7.0.0)
Requirement already satisfied: pytest-xdist in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (3.8.0)
Requirement already satisfied: exceptiongroup>=1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest) (1.2.2)
Requirement already satisfied: iniconfig>=1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest) (2.1.0)
Requirement already satisfied: packaging>=20 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest) (25.0)
Requirement already satisfied: pluggy<2,>=1.5 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest) (1.6.0)
Requirement already satisfied: pygments>=2.7.2 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest) (2.18.0)
Requirement already satisfied: tomli>=1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest) (2.2.1)
Requirement already satisfied: coverage>=7.10.6 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from coverage[toml]>=7.10.6->pytest-cov) (7.12.0)
Requirement already satisfied: execnet>=2.1 in /home/jovyan/.conda/envs/python-3.9/lib/python3.10/site-packages (from pytest-xdist) (2.1.1)
ImportError while loading conftest '/home/jovyan/intel-xpu-backend-for-triton/vllm/tests/conftest.py'.
tests/conftest.py:48: in <module>
    from tests.models.utils import TokensTextLogprobs, TokensTextLogprobsPromptLogprobs
tests/models/utils.py:13: in <module>
    from vllm.config.model import ModelConfig, ModelDType, RunnerOption
vllm/config/__init__.py:6: in <module>
    from vllm.config.compilation import (
vllm/config/compilation.py:23: in <module>
    from vllm.platforms import current_platform
vllm/platforms/__init__.py:256: in __getattr__
    _current_platform = resolve_obj_by_qualname(platform_cls_qualname)()
vllm/utils/import_utils.py:122: in resolve_obj_by_qualname
    module = importlib.import_module(module_name)
vllm/platforms/xpu.py:9: in <module>
    import vllm_xpu_kernels._C  # noqa
E   ModuleNotFoundError: No module named 'vllm_xpu_kernels._C'
```

--- a/scripts/test-triton.sh
+++ b/scripts/test-triton.sh
@@ -724,11 +724,19 @@ run_sglang_install() {
   echo "******    Installing SGLang                 ****"
   echo "************************************************"
 
-  if ! [ -d "./sglang" ]; then
-    git clone https://github.com/sgl-project/sglang.git
+  if pip show sglang >/dev/null 2>&1; then
+    echo "WARNING: sglang is already installed, skipping installation."
+    echo "To get clean installation, run:"
+    echo "  rm -rf ./sglang && pip uninstall -y sglang"
+    return
   fi
 
-  if ! pip list | grep "sglang" ; then
+  if [ -d "./sglang" ]; then
+    echo "WARNING: ./sglang directory already exists, installing from it."
+    echo "To get clean installation, run:"
+    echo "  rm -rf ./sglang && pip uninstall -y sglang"
+  else
+    git clone https://github.com/sgl-project/sglang.git
     cd sglang
     git checkout "$(<../benchmarks/third_party/sglang/sglang-pin.txt)"
     git apply ../benchmarks/third_party/sglang/sglang-test-fix.patch
@@ -741,9 +749,10 @@ run_sglang_install() {
     # We remove timm because it depends on torchvision, which depends on torch==2.9
     sed -i '/pytorch\|torch\|sgl-kernel\|timm/d' python/pyproject.toml
     cat python/pyproject.toml
-    pip install -e "./python"
     cd ..
   fi
+
+  pip install -e "./sglang/python"
 }
 
 run_sglang_tests() {
@@ -792,9 +801,35 @@ run_vllm_install() {
   echo "************************************************"
   echo "******    Installing VLLM                 ******"
   echo "************************************************"
+  CLEAN_MSG="To get a clean install, run: \n    rm -rf ./vllm ./vllm-xpu-kernels && pip uninstall -y vllm vllm-xpu-kernels"
 
-  if ! [ -d "./vllm" ]; then
+  local has_vllm_pip=false
+  local has_kernels_pip=false
+  pip show vllm >/dev/null 2>&1 && has_vllm_pip=true
+  pip show vllm-xpu-kernels >/dev/null 2>&1 && has_kernels_pip=true
+
+  # Both libraries already installed — nothing to do
+  if [ "$has_vllm_pip" = true ] && [ "$has_kernels_pip" = true ]; then
+    echo "WARNING: vllm and vllm-xpu-kernels are already installed, skipping installation."
+    echo -e $CLEAN_MSG
+    return
+  fi
+
+  # Only one of the two libraries installed — partial state, refuse to continue
+  if [ "$has_vllm_pip" = true ] || [ "$has_kernels_pip" = true ]; then
+    echo "ERROR: Partial vllm installation detected (vllm=$has_vllm_pip, vllm-xpu-kernels=$has_kernels_pip)."
+    echo -e $CLEAN_MSG
+    return 1
+  fi
+
+  # Neither library is installed — proceed, reusing existing directories if present
+  if [ -d "./vllm" ]; then
+    echo "WARNING: ./vllm directory already exists, installing from it."
+    echo -e $CLEAN_MSG
+  else
     git clone https://github.com/vllm-project/vllm.git
+
+    # Checkout the pinned commit, apply necessary patches and modify tests to run on xpu
     cd vllm
     git checkout "$(<../benchmarks/third_party/vllm/vllm-pin.txt)"
     git apply ../benchmarks/third_party/vllm/vllm-fix.patch
@@ -807,23 +842,26 @@ run_vllm_install() {
       tests/kernels/attention/test_triton_unified_attention.py
 
     cd ..
-    cp -r vllm/tests benchmarks/third_party/vllm/tests
   fi
+  # These files are neceassary for benchmarking runs
+  cp -r vllm/tests benchmarks/third_party/vllm/tests
 
-  if ! pip list | grep "vllm" ; then
-    pip install -r vllm/requirements/xpu.txt
+  pip install -r vllm/requirements/xpu.txt
 
+  if [ -d "./vllm-xpu-kernels" ]; then
+    echo "WARNING: ./vllm-xpu-kernels directory already exists, installing from it."
+    echo $CLEAN_MSG
+  else
     git clone https://github.com/vllm-project/vllm-xpu-kernels
     cd vllm-xpu-kernels
     git checkout "$(<../benchmarks/third_party/vllm/vllm-kernels-pin.txt)"
     sed -i '/pytorch\|torch/d' requirements.txt
     sed -i '/pytorch\|torch/d' pyproject.toml
     pip install -r requirements.txt
-    VLLM_TARGET_DEVICE=xpu pip install --no-build-isolation -e .
     cd ..
-
-    VLLM_TARGET_DEVICE=xpu pip install --no-deps --no-build-isolation -e vllm
   fi
+  VLLM_TARGET_DEVICE=xpu pip install --no-build-isolation -e vllm-xpu-kernels
+  VLLM_TARGET_DEVICE=xpu pip install --no-deps --no-build-isolation -e vllm
 
   pip install cachetools cbor2 blake3 pybase64 openai_harmony tblib
 }

